Fix: Sidebar becomes blank when navigating to reports or reloading pages.

Altrix-One opened the pull request:
Fixes an issue where the sidebar becomes blank when navigating to reports or reloading pages that aren't linked to workspaces. The fix adds proper null checks, URL decoding, and fallback workspace logic to ensure the sidebar is always displayed.

When navigating to reports (especially via direct URL) or reloading pages, the sidebar would become blank if:
- The report/page is not linked to any workspace
- The route is null/undefined/empty  
- The report name is URL-encoded
- No fallback workspace is available

This resulted in a poor user experience where users would lose their sidebar context.

<!--

Root Causes

1. Null pointer errors: Accessing `route[0]` without checking if route is null/undefined/empty
2. Missing fallback logic: When no workspace is found for a report, the code would return early without setting up any sidebar
3. URL encoding: Report names in URLs are URL-encoded (e.g., `Utilisation%20Trend%20Report`) but weren't being decoded before lookup
4. No preservation: When navigating to unhandled routes, the current sidebar wasn't preserved


Fixes issue where sidebar becomes blank when:
- Navigating to reports
- Reloading pages
- Routes are null/undefined/empty


<img width="1469" height="401" alt="image" src="https://github.com/user-attachments/assets/4c4b6b06-c2b4-425d-bd13-59f329425204" />

Altrix-One committed:
Fix: Sidebar becomes blank when navigating to reports or reloading pages

Fixes an issue where the sidebar becomes blank when navigating to reports or reloading pages that aren't linked to workspaces. The fix adds proper null checks, URL decoding, and fallback workspace logic to ensure the sidebar is always displayed.

When navigating to reports (especially via direct URL) or reloading pages, the sidebar would become blank if:
- The report/page is not linked to any workspace
- The route is null/undefined/empty  
- The report name is URL-encoded
- No fallback workspace is available

This resulted in a poor user experience where users would lose their sidebar context.

Altrix-One commented:
@sokumon This will be a value add.

sokumon commented:
@Altrix-One you had opened one more PR and now that is closed is that issue resolved? 

I am working on a fix which tries to semantically show the correct sidebar. Which shall resolve the issue of the empty sidebar

Altrix-One commented:
Yes the previous fix was for the old develop branch before Beta was merged. This was a new fix. But I trust your solution, end goal is to get a long term solution.

Altrix-One committed:
Feat: Refactored the sidebar resolution logic to be more semantic and aligned with Frappe patterns.

Improvements:

Semantic method names: Methods clearly describe their purpose:
- is_valid_route() - validates route structure
- resolve_workspace_for_route() - main resolution entry point
- resolve_workspace_for_doctype() - handles List/Form routes
- resolve_workspace_for_report() - handles report routes
- resolve_workspace_for_page() - handles page routes
- get_fallback_workspace() - unified fallback logic
- preserve_current_sidebar() - preserves sidebar state

Clear resolution strategy hierarchy:
- Priority 1: Explicit sidebar links
- Priority 2: Preserve current workspace if it matches
- Priority 3: Module-based resolution
- Priority 4: Default workspace from boot data
- Priority 5: First available workspace

Better code organization:
- Single responsibility per method
- JSDoc comments for documentation
- Switch statement for route type handling
- Consistent error handling

Maintainability:
- Easier to understand and modify
- Follows Frappe's coding conventions
- Reduces nested conditionals
- Clear separation of concerns

Features:
- Preserves current sidebar when appropriate
- Handles invalid routes gracefully
- URL-decodes report names
- Module-based fallback resolution
- Works with all route types (List, Form, Report, Page, Workspaces)

The code is now more maintainable and follows Frappe's semantic approach to sidebar resolution.

Altrix-One commented:
@sokumon  @umairsy 

Refactored `set_workspace_sidebar()` method to use a semantic, strategy-based approach for workspace resolution. Transforms complex nested conditionals into clear, maintainable methods that follow Frappe coding standards.

Refactored Methods

- `set_workspace_sidebar(router)`: Main entry point - validates route and delegates to route-specific resolvers
- `resolve_workspace_for_route(route, router)`: Route type dispatcher using switch statement
- `resolve_workspace_for_doctype(route, router)`: Handles List/Form routes with priority-based resolution
- `resolve_workspace_for_report(route, router)`: Handles report routes with URL decoding
- `resolve_workspace_for_page(route, router)`: Handles page routes and unhandled route types
- `get_fallback_workspace(router)`: Unified fallback strategy with priority hierarchy

New Helper Methods

- `is_valid_route(route)`: Validates route structure
- `normalize_report_name(report_name)`: Safely decodes URL-encoded report names
- `should_preserve_current_workspace(sidebars)`: Determines if current workspace should be preserved
- `select_best_workspace(sidebars)`: Selects optimal workspace from available options
- `resolve_workspace_by_module(router)`: Module-based workspace resolution
- `preserve_current_sidebar()`: Preserves sidebar state on errors


This has been tested and works all over :

Reports now keep the correct sidebar:
<img width="350" height="441" alt="image" src="https://github.com/user-attachments/assets/c975cb8b-7fda-45f0-8873-20a4340aa4a1" />

Pages can be refreshed and sidebar stays intact:
<img width="586" height="612" alt="image" src="https://github.com/user-attachments/assets/37d6ae0c-229e-4df0-9436-68f1d51d5726" />

Normal Doctypes are unchanged:
<img width="370" height="368" alt="image" src="https://github.com/user-attachments/assets/0a117a99-f0e1-4a40-b165-285f92f17189" />


This has been thoroughly tested, lets please get this in.

sokumon commented:
I like the approach you have taken but it will take me some time to review it throughtly as I am working on something. I will get back to you

In the mean time can you fix the other issues 

- [ ] Run Pre-commit
- [ ] Do semantic commit (Change Fix to fix)

Altrix-One committed:
fix: pre-commit and semantic commit

Altrix-One commented:
@sokumon I have applied the needed changes.

sokumon commented on frappe/public/js/frappe/ui/sidebar/sidebar.js:
```diff
@@ -406,7 +604,6 @@ frappe.ui.Sidebar = class Sidebar {
 	toggle_editing_mode() {
 		const me = this;
 		if (this.edit_mode) {
-			this.open();
```
I had added this to force open when edit sidebar is clicked

Please add it back

Altrix-One committed:
fix: re-introduce function that when sidebar is clicked, the sidebar is forced open

Altrix-One commented:
@sokumon I have re-introduced your code. Sorry for removing it.

Altrix-One commented:
@sokumon let me know what is still needed here.

sokumon commented:
Hello @Altrix-One 
Although this was good idea to avoid showing an empty sidebar. The better idea is to generate sidebar for all modules https://github.com/frappe/frappe/pull/34918. and then render the correct sidebar for the page. 

Thanks for the contribution :)
